### ISSUE TITLE
documentation viewer

### ISSUE THREAD

**Issue by LecrisUT** [catchorg/Catch2#2709]
**Description**
Currently there are 2 open PRs suggesting documentation builders:
- #2672 
- #2478 
- #2710 

I believe that is a valuable feature, but I believe it's missing RTD suppott that would make the documentations:
- searchable
- version specific
- viewable on PRs

For this I think it's valuable to implement an established standard of sphinx. What are your opinions?

**Comment by adehad** [comment]
Sphinx with the readthedocs github integration (so it runs on each PR etc) would be a great option.

Happy to support this effort, I have got it working with markdown files before

**Comment by nrbnlulu** [comment]
IMO mkdocs-material looks much better then Sphinx although Sphinx is considered more mature, AFAIK recent projects
tends to use mkdocs over sphinx. 

> searchable - :heavy_check_mark:   

![image](https://github.com/catchorg/Catch2/assets/88795475/e3621afd-bad7-4dc4-ae8e-a6cbb7d95403)

> version specific  

[Doable on mkdocs](https://squidfunk.github.io/mkdocs-material/setup/setting-up-versioning/)

> viewable on PRs

This is not specific to mkdocs or Sphinx, You'll need to setup a PaaS for that, I don't think github
support this out of the box.

**Comment by LecrisUT** [comment]
Both mkdocs and sphinx are supported for rtd out-of-the-box, so I am inclined to either.

On the specifics of mkdocs vs sphinx, I want to check:
- Are there equivalents to [myst-parser](https://myst-parser.readthedocs.io/en/latest/syntax/admonitions.html) supported functions, specifically:
  - Admotions like caution, warning, note, etc.
  - Tab-set
  - Include from-to functionality, either as code or as markdown to be processed
 - Is there a set of themes to see them in action like at sphinx-themes.org
 - Is there support for doxygen like in breathe? E.g. including contents of a whole doxygen group?
 - Does mkdocs have an autobuild/live editor where when you change the md source or doxygen comments it rebuilds of itself so you have a live web view of it?
 - Are there any other specific features to mkdocs that might not be on sphinx?
 - This one is a bonus, is there an equivalent to [inter-sphinx](https://myst-parser.readthedocs.io/en/latest/syntax/cross-referencing.html#cross-project-intersphinx-links), where you can link to items in other projects and you get a pop-up with a preview of that content?

**Comment by nrbnlulu** [comment]
> * Are there equivalents to [myst-parser](https://myst-parser.readthedocs.io/en/latest/syntax/admonitions.html) supported functions, specifically:
>   
>   * Admotions like caution, warning, note, etc. - :heavy_check_mark: 
>   * Tab-set - :heavy_check_mark: 
>   * Include from-to functionality, either as code or as markdown to be processed - :heavy_check_mark: 
> * Is there a set of themes to see them in action like at sphinx-themes.org - :x: 
Everyone just use mkdocs-material because it looks awesome, well maintained, and feature rich. 
> * Is there support for doxygen like in breathe? E.g. including contents of a whole doxygen group?
> * Does mkdocs have an autobuild/live editor where when you change the md source or doxygen comments it rebuilds of itself so you have a live web view of it?

No, (not that sphinx does) we'll need to craft something on our own using the xml doxygen generates...
> * Are there any other specific features to mkdocs that might not be on sphinx?

I don't think so.
> * This one is a bonus, is there an equivalent to [inter-sphinx](https://myst-parser.readthedocs.io/en/latest/syntax/cross-referencing.html#cross-project-intersphinx-links), where you can link to items in other projects and you get a pop-up with a preview of that content?

Not that I know.
___
Note that the majority of features that Sphinx provide over mkdocs are Python specific. 
I only use mkdocs to generate html from markdown and it does it much better then Sphinx IMO.

**Comment by LecrisUT** [comment]
Not having `breathe` is a major bummer for a cpp project. There are quite a few api documentations that should br exposed and synchronized between the docs file and in-code.

It is helpful to have this discussion before we get locked-in by designing for a specific ecosystem. If the only criteria is the looks, then check the sphinx-theme.org and comment on what is missing there. I like RTD and furo theme. Other contenders are book, pydata and materials (lesser because of how api view is generated out-of-the-box).

Both ecosystems are constantly growing so who knows how they will evolve.

**Comment by nrbnlulu** [comment]
> Not having `breathe` is a major bummer for a cpp project. There are quite a few api documentations that should br exposed and synchronized between the docs file and in-code.

Typical Catch2 usage is with macros more than with classes so it is not much of a bummer + it doesn't look very good and we can craft something ourselfs using doxygen xml.  

> If the only criteria is the looks, then check the sphinx-theme.org and comment on what is missing there. I like RTD and furo theme. Other contenders are book, pydata and materials (lesser because of how api view is generated out-of-the-box).

- mkdocs-material is incomparable to these themes and is much more mature.

- It is not the only one, mkdocs is much easier to understand, maintain, extend and it is build for markdown while sphinx is 
first-class rst, although it has markdown integrations.

**Comment by LecrisUT** [comment]
> Typical Catch2 usage is with macros more than with classes so it is not much of a bummer + it doesn't look very good

Note that macros are supported by doxygen and breathe. [Source](https://github.com/breathe-doc/breathe/blob/bf020a8c284929f48033c0c8fb9de37615d5c0c1/examples/specific/define.h) and [generated](https://breathe.readthedocs.io/en/latest/specific.html#id1).

Also with C++ modules, it is unclear how macros will continue to exist as exported API. It will be great if the api behind the macros are also documented.

> sphinx is first-class rst, although it has markdown integrations.

`myst-parser` is plenty mature at this point, and various projects rely on it.

**Comment by nrbnlulu** [comment]
Yeah, anyways, thanks for your input! I would continue my mkdocs PR tonight :crossed_fingers: and leave for maintainers
to decide what would be best...

**Comment by nrbnlulu** [comment]
FWIW here is mkdocs C++ generated API https://json.nlohmann.me/api/basic_json/

**Comment by hwhsu1231** [comment]
Any progress of this issue? Personally, I prefer the Sphinx one, since the layout is more beautiful to me.

Screenshot from @LecrisUT 's example: https://catch2-temp.readthedocs.io/en/latest/

![image](https://github.com/catchorg/Catch2/assets/71438617/f04377df-b1dd-49dd-baab-e9f13c5a23bd)
